File: Problem1.java
// # BFS-1
// # Problem 1
// Binary Tree Level Order Traversal (https://leetcode.com/problems/binary-tree-level-order-traversal/)

// Time Complexity : BFS with Queue - O(N)
// Space Complexity : BFS with Queue - O(N/2) max for leaf nodes in worst case ==> O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach in three sentences only


class Solution {
    // 1. Use a queue to do BFS
    // Time: O(N)
    // Space: O(N/2) max for leaf nodes in worst case ==> O(N)
    public List<List<Integer>> bfsHelper(TreeNode root) {
        Queue<TreeNode> q=new LinkedList<>();

        // Result
        List<List<Integer>> result=new ArrayList<>();
        if(root==null){
            return result;
        }

        // Add root node
        q.add(root);
        // Size at level n
        int size_n;

        while(!q.isEmpty()){
            size_n=q.size();

            // Process current level
            List<Integer> currLevel= new ArrayList<>();
            for(int i=0;i<size_n;i++){
                TreeNode curr=q.remove();
                
                // Process the current node
                currLevel.add(curr.val);

                // Add children of curr node to queue
                if(curr.left!=null){
                    q.add(curr.left);
                }
                if(curr.right!=null){
                    q.add(curr.right);
                }

            }

            // Add curr level list to result
            result.add(currLevel);
        }

        return result;
    }


    private void dfs(TreeNode root, int level, Map<Integer, List<Integer>> levelMap){
        if(root==null)  return;

        // Process current node
        if(!levelMap.containsKey(level)){
            levelMap.put(level, new ArrayList<>());
        }

        levelMap.get(level).add(root.val);

        dfs(root.left, level+1, levelMap);
        dfs(root.right, level+1, levelMap);
        
    }

    // 2. Maintain a hasmap of levels -> nodes
    // At each call maintain current level as parameter
    // Iterate over the map and store in result
    // Time: O(N)
    // Space: O(H), recursive stack space
    public List<List<Integer>> dfsHelper(TreeNode root){
        List<List<Integer>> result = new ArrayList<>();
        Map<Integer, List<Integer>> levelMap=new HashMap<>();

        // dfs using inOrder
        dfs(root, 0, levelMap);

        // Iterate over map from 0 till curr level doesn't exist
        int currLevel=0;
        while(levelMap.containsKey(currLevel)){
            result.add(levelMap.get(currLevel));
            currLevel++;
        }

        return result;
    }

    private void dfsWithList(TreeNode root, int level, List<List<Integer>> result){
        //Base
        if(root==null){
            return;
        }

        //Logic
        while(result.size()<=level){
            result.add(new ArrayList<>());
        }
        result.get(level).add(root.val);
        dfsWithList(root.left,level+1,result);
        dfsWithList(root.right,level+1,result);

    }

    // 3. Maintain a list of list of nodes, index of first list corresponds to it's level
    // At each call maintain current level as parameter
    // Time: O(N)
    // Space: O(H), recursive stack space
    public List<List<Integer>> dfsWithListHelper(TreeNode root){
        List<List<Integer>> result = new ArrayList<>();

        // dfs using inOrder
        dfsWithList(root, 0, result);
        return result;
    }

    

    public List<List<Integer>> levelOrder(TreeNode root){
        //1. Using BFS Queue
        //return bfsHelper(root);
        //2. DFS and hashmap
        //return dfsHelper(root);
        //3. DFS with list
        return dfsWithListHelper(root);
    }
}

File: Problem_1.py
#
# @lc app=leetcode id=102 lang=python3
#
# [102] Binary Tree Level Order Traversal
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
'''
Time Complexity - O(n). We traverse each and every node of the Tree for DFS and O(n) + O(h) for DFS
Space Complexity - O(n). We are using a queue where at the worst case we will have (n/2) nodes i.e. nodes at last level.
                   O(n) for DFS as well.

This code works on Leetcode
'''
from collections import deque
class Solution:
    def __init__(self):
        self.levelMap = {}

    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        result = []
        # USING BFS
        nodeQ = deque() #create Queue
        nodeQ.append(root) #add the root to the queue
        while nodeQ:
            qlen = len(nodeQ) #before processing the queue find size of queue. We are doing this to distinguish between levels
            level = [] #create a list for each level
            for i in range(qlen):
                currNode = nodeQ.popleft() #takeout the first element from the queue for processing
                if currNode: #if node is not null 
                    level.append(currNode.val) #add the value to the level list
                    nodeQ.append(currNode.left) #add the left children to the list first
                    nodeQ.append(currNode.right) #Then add all the right children to the list
            if level: #if we have a node in the list, append the list to the result
                result.append(level)
        # USING DFS
        # self.helper(root, 0)
        # lvl = 0
        # while lvl in self.levelMap: #for every level get the nodes until we find a level that does not exist
        #     result.append(self.levelMap.get(lvl)) #add all the nodes to the result
        #     lvl+=1 #increment level to access next nodes
        return result #finally return the result
    
    def helper(self, root, currLvl):
        if root == None: #return if root is null
            return
        if currLvl not in self.levelMap: #if this is the first element at that particular level
            self.levelMap[currLvl] = []
        lvlNodes = self.levelMap.get(currLvl) 
        lvlNodes.append(root.val) #else add that element to the corresponding level
        self.helper(root.left, currLvl+1) #traverse left, going down by 1 level hence current level + 1
        self.helper(root.right, currLvl+1) #traverse right, going down by 1 level hence current level + 1

    
        
# @lc code=end

